keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,759 @@
+- put pods with consecutive indexes of an IndexedJob on close nodes
+- support the following features: reclaimable, pods, cohorts preemption
+ within cluster queue
+- support re-computing the TopologyAssignment while the workload
+ is running
+- perform full scheduling simulation rather than just capacity counting
+ (including pod affinities and anti-affinities)
+
+## Implementation History
+
+- 2024-07-25 - PR merged [Allow mutating schedulingGates when the Jobset is suspended](Allow mutating schedulingGates when the Jobset is suspended)
+- 2024-07-30 - this KEP
+
+
+## Drawbacks
+
+Tracking of nodes by Kueue will increase its code complexity and memory
+consumption.
+
+## Alternatives
+
+### Account usage by watching DaemonSet pods
+
+The idea is to account for DaemonSet pods by tracking the DaemonSets rather than
+their Pods. This can prevent the [race condition](Race condition when accounting
+for DaemonSet pods) risk.
+
+**Reasons for discarding/deferring**
+
+When accounting for the the DaemonSets we would need to verify their selectors
+would allow match a node in question.
+
+This is a complication which may not be necessary for MVP since the workloads
+will mostly expected to consume accelerators (GPUs and TPUs).
+
+We will consider this for Beta or GA based on the users' feedback.
+
+### Use label for workload
+
+We considered using the `kueue.x-k8s.io/workload` label rather than annotation.
+
+**Reasons for discarding/deferring**
+
+Currently workloads can have names which exceed the maximal length for a label.
+So, we would need to shorten the maximal workload name by reducing the const
+`maxPrefixLength` in `workload_names.go`. However, we can also facilitate
+fast lookups based on labels if we index the workloads.
+
+### Implement it in ClusterAutoscaler or kube-scheduler
+
+This feature requires tracking of the nodes to reconstruct the nodes capacity
+for all topology domains, and this is a novelty in Kueue which didn't do that,
+leaving the features requiring knowledge about Nodes to kube-scheduler or
+Cluster Autoscaler.
+
+**Reasons for discarding/deferring**
+
+kube-scheduler:
+- scheduler does not have ability to do what `TopologyUngater` does, it is
+  lower-level of abstraction
+- the topology information is not represented in the core k8s, and scheduler
+  wouldn't read the information from CRDs
+- kube-scheduler doesn't do PodGroup level scheduling or preemption which is
+  required
+
+ClusterAutoscaler:
+- Isn't focused on scheduling or preemption; its responsibility is to provision
+  new Nodes
+- Operates at the level of individual Pods rather than Jobs or PodGroups
+
+Also, both of the components don't have a concept of quota, which we want to be
+supported with this feature. So, if Kueue admits a Job with "required: rack"
+just based on quota, the Pods might be created, but the components wouldn't
+be able to schedule the Pods if there is not enough capacity in the topology
+domain.
```

Comment:
They do have a retry mechanism. AFAIK the components don't have a native concept of Pod group. You can simulate Pod group using Volcano in kube-scheduler, but kube-scheduler won't be able to provision the nodes by itself if they don't exist.

Reply:
That makes sense. Maybe we can delegate TAS scheduling to kube-scheduler after we support the native PodGroup feature: https://github.com/kubernetes/enhancements/issues/4671

But, there are not progressing in the k/k now.

Reply:
Even then I think there are other blockers like lack of API to support topology. I think it makes sense to incubate such an API in CRDs before moving to the core.

Reply:
Yes, I agree with you.
As an alternative, we can provide the kube-scheduler-plugin for TAS in Kueue side instead of watching Node resources in the kueue, but I guess the kueue custom kube-scheduler-plugin may bring batch admins obstacles to install Kueue for batch admins.

So, I add +1 to this node watch mechanism.

But, if we add more and more features depending on the node information, I'm wondering if we should revisit here and discuss if we should provide the custom kube-scheduler plugin for Kueue.